README.rst
Thoth Core Management API
=========================

.. image:: https://img.shields.io/github/v/tag/thoth-station/management-api?style=plastic
  :target: https://github.com/thoth-station/management-api/releases
  :alt: GitHub tag (latest by date)

.. image:: https://quay.io/repository/thoth-station/management-api/status
  :target: https://quay.io/repository/thoth-station/management-api?tab=tags
  :alt: Quay - Build

.. image:: https://api.codacy.com/project/badge/Grade/d8f62cde59b84854ac425d148570f1ab
   :alt: Codacy Badge
   :target: https://app.codacy.com/app/thoth-station/management-api?utm_source=github.com&utm_medium=referral&utm_content=thoth-station/management-api&utm_campaign=Badge_Grade_Dashboard

This API service is used for administrative and operational tasks for a Thoth
deployment. For Management API interaction, one needs a token that can be
obtained by contacting Thoth deployment administrator.

Installation and deployment
###########################

The service is built using OpenShift Source-to-Image and deployed
automatically via Argo CD - see `thoth-station/thoth-application
repository <https://github.com/thoth-station/thoth-application>`_.

Running Management API locally
##############################

Management API can be run locally in a mode when it still talks to the cluster.
To run Management API locally, create ``.env`` file out out ``.env.template``
and adjust environment variable values as desired (see `thoth-station/storages
<https://github.com/thoth-station/storages/>`__ for more info):

.. code-block:: console

  cp .env.template .env
  vim .env

Once the environment is properly setup, you can run Management API locally:

.. code-block:: console

  pipenv install
  pipenv run gunicorn thoth.management_api.openapi_server:app --config gunicorn.conf.py


CHANGELOG.md
# Changelog for Thoth's Management API Service

## Release 0.18.11 (2023-01-17T21:54:02)
* 0cf0021 ':arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment' (#945)

## Release 0.18.10 (2022-12-05T04:37:55)
* d2603b1 Fix the DatabaseNotInitializedError exception from thoth-storages (#941)

## Release 0.18.9 (2022-12-05T01:58:58)
* e23bbab ':arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment'
* e6ad630 ':arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment' (#935)
* d331ee6 ':arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment' (#934)
* 2002ac4 :ship: Bump up base image initialized in CI. (#933)
* 1a27cd2 git ls-files -z -- .pre-commit-config.yaml | xargs -0 sed -i 's#https://gitlab.com/PyCQA/flake8#https://github.com/PyCQA/flake8#' (#932)
* 9a6bde5 :ship: Bump up base image initialized in CI.

## Release 0.18.8 (2022-10-13T18:55:29)
* 422bd1b :recycle: Correct the error type for exception handle in get_log
* bf7500a :books: Adjusted the return with description and .env.template for local dev
* 428fb28 :hammer: Remove the jaegar tracing from the api
* 0ab52d9 :guardsman: Update the OWNERS and pre-commit hooks
* f2040db :ship: Bump up base image initialized in CI.

## Release 0.18.7 (2022-09-20T02:04:42)
* e99e968 :ship: Bump up base image initialized in CI. (#924)
* b84290d :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment (#923)
* b77e746 :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment (#921)
* 9bb95ad :ship: Bump up base image initialized in CI.
* 4e4247e :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment

## Release 0.18.6 (2022-08-23T04:08:21)
* 3da0734 :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment (#914)
* b3f8315 :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment (#913)
* edbb166 Fix a typo in ci config (#910)
* 91bdad3 :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment
* e6f03cf :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment (#909)
* 8d91ad5 :ship: Bump up base image initialized in CI. (#908)
* 9d56d32 Get flake8 config out of pre-commit config (#907)

## Release 0.18.5 (2022-07-07T13:41:05)
* 4c8fba8 Remove ref to external doc and explicit the parameter purpose
* e6233f4 Fix the description of index_url in /python/package-index/state
* 7ec2955 :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment
* b10bc18 Fix: To check if 'warehouse_api_url' key exists (#897)
* 63d1d3b :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment
* bc61508 enable sync-job to be called from the management api (#884)
* c8d6860 :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment

## Release 0.18.4 (2022-05-09T19:29:30)
* 70a0219 Fix retrieving unsolvable packages (#890)
* 195a29a :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment
* ab54c49 :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment (#888)
* 13b98e9 :ship: Bump up base image initialized in CI. (#886)
* 7b926f9 :ship: Bump up base image initialized in CI.
* 0f9337f :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment
* e3faa44 Update url to amun-api prod
* 61b4656 :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment
* 7525596 Add my username to approvers and maintainers

## Release 0.18.3 (2022-04-07T11:26:03)
* fab7f87 Remove requirements_locked and constraints from dependency monkey

## Release 0.18.2 (2022-04-07T08:41:45)
* a0497d2 Downgrade markupsafe <2.1.0

## Release 0.18.1 (2022-04-06T14:00:51)
* 8492f47 :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment
* bc244f3 Keep base image to latest tag
* 50221a7 Update pre-commit base image version
* 8acb248 Update dependency-monkey endpoint schema
* 88d2bc7 :ship: Bump up base image initialized in CI.
* 087983e Fix pre-commit git protocol and config
* 3612a38 :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment
* 3b01329 Document how to run Management API locally
* 1336ac1 :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment

## Release 0.18.0 (2022-02-14T09:41:43)
* :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment
* Allow passing dependency indexes to the solver workflow

## Release 0.17.16 (2022-01-31T21:57:56)
* :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment
* Remove unused .thothTemplate.yaml
* :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment

## Release 0.17.15 (2021-12-21T22:09:01)
* :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment
* :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment
* :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment

## Release 0.17.14 (2021-11-08T12:25:59)
* :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment

## Release 0.17.13 (2021-10-18T12:17:28)
* Adjust parameter in method
* :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment

## Release 0.17.12 (2021-10-15T21:05:05)
* :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment

## Release 0.17.11 (2021-09-27T21:46:43)
### Features
* :arrow_up: Automatic update of dependencies by Kebechet for the ubi8 environment

## Release 0.17.10 (2021-09-14T20:54:51)
### Features
* Update of the dependencies
* :arrow_up: Automatic update of dependencies by Kebechet

## Release 0.17.9 (2021-08-11T11:43:44)
### Features
* Make warehouse_api_url optional

## [0.0.1] - 2018-Jun-26 - fpokorny

### Added

* Initial Management API implementation



thoth/management_api/entrypoint.py
#!/usr/bin/env python3
# thoth-management-api
# Copyright(C) 2018, 2019 Fridolin Pokorny
#
# This program is free software: you can redistribute it and / or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
"""Core Thoth Management API."""

import logging
from datetime import datetime

import connexion

from flask import redirect, request, jsonify
from flask_script import Manager
from prometheus_flask_exporter import PrometheusMetrics

from thoth.common import SafeJSONEncoder
from thoth.common import init_logging
from thoth.common import logger_setup
from thoth.storages import SolverResultsStore

import thoth.management_api as thoth_management_api

from .configuration import Configuration


init_logging()
_LOGGER = logging.getLogger("thoth.management_api")

# Expose for uWSGI.
app = connexion.App(__name__)
application = app.app
metrics = PrometheusMetrics(application)

app.add_api(Configuration.SWAGGER_YAML_PATH)
application.json_encoder = SafeJSONEncoder
manager = Manager(application)

# Needed for session.
application.secret_key = Configuration.APP_SECRET_KEY

# static information as metric
metrics.info(
    "management_api_info",
    "Management API info",
    version=thoth_management_api.__version__,
)


@app.route("/")
def base_url():
    """Redirect to UI by default."""
    return redirect("api/v1/ui")


@app.route("/api/v1")
@metrics.do_not_track()
def api_v1():
    """Provide a listing of all available endpoints."""
    paths = []

    for rule in application.url_map.iter_rules():
        rule = str(rule)
        if rule.startswith("/api/v1"):
            paths.append(rule)

    return jsonify({"paths": paths})


@metrics.do_not_track()
def _healthiness():
    """Check service healthiness."""
    # Check that Ceph is reachable.
    adapter = SolverResultsStore()
    adapter.connect()
    adapter.ceph.check_connection()

    return (
        jsonify({"status": "ready", "version": thoth_management_api.__version__}),
        200,
        {"ContentType": "application/json"},
    )


@logger_setup("werkzeug", logging.WARNING)
@logger_setup(
    "botocore.vendored.requests.packages.urllib3.connectionpool", logging.WARNING
)
@app.route("/readiness")
@metrics.do_not_track()
def api_readiness():
    """Report readiness for OpenShift readiness probe."""
    return _healthiness()


@logger_setup("werkzeug", logging.WARNING)
@logger_setup(
    "botocore.vendored.requests.packages.urllib3.connectionpool", logging.WARNING
)
@app.route("/liveness")
@metrics.do_not_track()
def api_liveness():
    """Report liveness for OpenShift readiness probe."""
    return _healthiness()


@application.errorhandler(404)
@metrics.do_not_track()
def page_not_found(exc):
    """Adjust 404 page to be consistent with errors reported back from API."""
    # Flask has a nice error message - reuse it.
    return jsonify({"error": str(exc)}), 404


@application.errorhandler(500)
@metrics.do_not_track()
def internal_server_error(exc):
    """Adjust 500 page to be consistent with errors reported back from API."""
    # Provide some additional information so we can easily find exceptions in logs (time and exception type).
    # Later we should remove exception type (for security reasons).
    return (
        jsonify(
            {
                "error": "Internal server error occurred, please contact administrator with provided details.",
                "details": {
                    "type": exc.__class__.__name__,
                    "datetime": datetime.utcnow().isoformat(),
                },
            }
        ),
        500,
    )


if __name__ == "__main__":
    _LOGGER.info(f"Thoth Management API v{thoth_management_api.__version__}")
    manager.run()


thoth/management_api/exceptions.py
#!/usr/bin/env python3
# thoth-management-api
# Copyright(C) 2018 - 2021 Fridolin Pokorny
#
# This program is free software: you can redistribute it and / or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

"""Exceptions raised in the whole management-api implementation."""


class ManagementApiException(Exception):  # noqa: N818
    """A base class for management API exceptions."""


class NotFoundException(ManagementApiException):  # noqa: N818
    """An exception raised if the requested resource could not be found."""


thoth/management_api/configuration.py
#!/usr/bin/env python3
# thoth-management-api
# Copyright(C) 2018 - 2021 Fridolin Pokorny
#
# This program is free software: you can redistribute it and / or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

"""Configuration of management API service."""

import logging
import os

_LOGGER = logging.getLogger(__name__)

_AMUN_API_URL = os.getenv("AMUN_API_URL") or "-"
if _AMUN_API_URL == "-":
    _LOGGER.error(
        "Amun API URL was not configured, Dependency Monkey results will not "
        "be submitted to Amun for inspection!"
    )


class Configuration:
    """Configuration of management API service."""

    APP_SECRET_KEY = os.environ["THOTH_APP_SECRET_KEY"]
    SWAGGER_YAML_PATH = os.path.join(
        os.path.dirname(os.path.realpath(__file__)), "../../openapi"
    )
    THOTH_DEPENDENCY_MONKEY_STACK_OUTPUT = _AMUN_API_URL
    THOTH_MANAGEMENT_API_TOKEN = os.environ["THOTH_MANAGEMENT_API_TOKEN"]
    THOTH_MIDDLETIER_NAMESPACE = os.environ["THOTH_MIDDLETIER_NAMESPACE"]
    THOTH_AMUN_INSPECTION_NAMESPACE = os.environ["THOTH_AMUN_INSPECTION_NAMESPACE"]
    THOTH_DEPLOYMENT_NAME = os.environ["THOTH_DEPLOYMENT_NAME"]

    OPENAPI_PORT = 8080


thoth/management_api/openapi_server.py
#!/usr/bin/env python3
# management-api
# Copyright(C) 2019 - 2021 Christoph Görn
#
# This program is free software: you can redistribute it and / or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

"""Thoth Management API entrypoint."""


import os
import sys
import logging
from datetime import datetime

import connexion
from connexion.resolver import RestyResolver

from flask import redirect, jsonify, request
from prometheus_flask_exporter import PrometheusMetrics

from thoth.common import __version__ as __common__version__
from thoth.storages import __version__ as __storages__version__
from thoth.common import datetime2datetime_str
from thoth.common import init_logging
from thoth.storages import GraphDatabase
from thoth.storages.exceptions import DatabaseNotInitializedError
from thoth.management_api import __version__
from thoth.management_api.configuration import Configuration

# Configure global application logging using Thoth's init_logging.
init_logging(logging_env_var_start="THOTH_MANAGEMENT_API_LOG_")

_LOGGER = logging.getLogger("thoth.management_api")
_LOGGER.setLevel(
    logging.DEBUG
    if bool(int(os.getenv("THOTH_MANAGEMENT_API_DEBUG", 0)))
    else logging.INFO
)

__service_version__ = (
    f"{__version__}+storage.{__storages__version__}.common.{__common__version__}"
)

_LOGGER.info("This is Management API v%s", __service_version__)
_LOGGER.debug("DEBUG mode is enabled!")

_THOTH_API_HTTPS = bool(int(os.getenv("THOTH_API_HTTPS", 1)))

# Expose for uWSGI.
app = connexion.FlaskApp(
    __name__, specification_dir=Configuration.SWAGGER_YAML_PATH, debug=True
)

app.add_api(
    "openapi.yaml",
    options={"swagger_ui": True},
    arguments={"title": "Management API"},
    resolver=RestyResolver("thoth.management_api"),
    strict_validation=True,
    validate_responses=bool(int(os.getenv("THOTH_API_VALIDATE_RESPONSES", 0))),
)


application = app.app

# create metrics and manager
metrics = PrometheusMetrics(application)

# Needed for session.
application.secret_key = Configuration.APP_SECRET_KEY

# static information as metric
metrics.info("management_api_info", "Management API info", version=__service_version__)
_API_GAUGE_METRIC = metrics.info(
    "management_api_schema_up2date", "Management API schema up2date"
)

# Instantiate one GraphDatabase adapter in the whole application (one per wsgi worker) to correctly
# reuse connection pooling from one instance.
GRAPH = GraphDatabase()

# custom metric to expose head revision from thoth-storages library
schema_revision_metric = metrics.info(
    "thoth_database_schema_revision_script",
    "Thoth database schema revision from script",
    component="management-api",  # label
    revision=GRAPH.get_script_alembic_version_head(),  # label
    env=Configuration.THOTH_DEPLOYMENT_NAME,  # label
)

GRAPH.connect()


@application.before_request
def before_request_callback():
    """Register callback, runs before each request to this service."""
    method = request.method
    path = request.path

    # Update up2date metric exposed.
    if method == "GET" and path == "/metrics":
        try:
            _API_GAUGE_METRIC.set(int(GRAPH.is_schema_up2date()))
        except DatabaseNotInitializedError as exc:
            # This can happen if database is erased after the service has been started as we
            # have passed readiness probe with this check.
            _LOGGER.exception(
                "Cannot determine database schema as database is not initialized: %s",
                str(exc),
            )
            _API_GAUGE_METRIC.set(0)


@application.before_first_request
def before_first_request_callback():
    """Register callback, runs before first request to this service."""
    schema_revision_metric.set(1)

    if bool(int(os.getenv("THOTH_MANAGEMENT_API_RUN_MIGRATIONS", 0))):
        GRAPH.initialize_schema()


@app.route("/")
@metrics.do_not_track()
def base_url():
    """Redirect to UI by default."""
    # https://github.com/pallets/flask/issues/773
    request.environ["wsgi.url_scheme"] = "https" if _THOTH_API_HTTPS else "http"
    return redirect("api/v1/ui/")


@app.route("/api/v1")
@metrics.do_not_track()
def api_v1():
    """Provide a listing of all available endpoints."""
    paths = []

    for rule in application.url_map.iter_rules():
        rule = str(rule)
        if rule.startswith("/api/v1"):
            paths.append(rule)

    return jsonify({"paths": paths})


def _healthiness():
    return (
        jsonify({"status": "ready", "version": __version__}),
        200,
        {"ContentType": "application/json"},
    )


@app.route("/readiness")
@metrics.do_not_track()
def api_readiness():
    """Report readiness for OpenShift readiness probe."""
    return _healthiness()


@app.route("/liveness")
@metrics.do_not_track()
def api_liveness():
    """Report liveness for OpenShift readiness probe."""
    return _healthiness()


@application.errorhandler(404)
@metrics.do_not_track()
def page_not_found(exc):
    """Adjust 404 page to be consistent with errors reported back from API."""
    # Flask has a nice error message - reuse it.
    return jsonify({"error": str(exc)}), 404


@application.errorhandler(500)
@metrics.do_not_track()
def internal_server_error(exc):
    """Adjust 500 page to be consistent with errors reported back from API."""
    # Provide some additional information so we can easily find exceptions in logs (time and exception type).
    # Later we should remove exception type (for security reasons).
    return (
        jsonify(
            {
                "error": "Internal server error occurred, please contact administrator with provided details.",
                "details": {
                    "type": exc.__class__.__name__,
                    "datetime": datetime2datetime_str(datetime.utcnow()),
                },
            }
        ),
        500,
    )


@application.after_request
def apply_headers(response):
    """Add headers to each response."""
    response.headers["X-Management-API-Version"] = __version__
    response.headers["X-Management-API-Service-Version"] = __service_version__
    return response


if __name__ == "__main__":
    app.run()

    Configuration.tracer.close()

    sys.exit(1)
